Update StringImpl::copyCharacters() to take spans and do bounds validation

cdumez opened the pull request:
#### 1ab6226d0714c683a76810fa340e29c48856faa1
<pre>
Update StringImpl::copyCharacters() to take spans and do bounds validation
<a href="https://bugs.webkit.org/show_bug.cgi?id=286435">https://bugs.webkit.org/show_bug.cgi?id=286435</a>

Reviewed by Darin Adler.

Also fix an issue in `JSRopeString::resolveToBuffer()` where we were artificially
reducing the size of the output buffer we were passing to `StringView::getCharacters()`
as it was tripping the subspan assertions. The code was safe since the buffer was large
enough, however, `StringView::getCharacters()` could not know that.

Note that the assertions are only debug assertions for now due to performance concerns
but we should try and make them release assertions wherever performance allows. The
assertions are currently in StringImpl::copyCharacters() but I intend to move them
to copyElements() in a follow-up as this requires even more refactoring.

* Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
(JSC::appendStringToData):
* Source/JavaScriptCore/runtime/StringConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::jsSpliceSubstrings):
* Source/WTF/wtf/HexNumber.h:
(WTF::StringTypeAdapter&lt;HexNumberBuffer&gt;::writeTo const):
* Source/WTF/wtf/URLHelpers.cpp:
(WTF::URLHelpers::escapeUnsafeCharacters):
* Source/WTF/wtf/text/StringBuilder.cpp:
(WTF::StringBuilder::append):
* Source/WTF/wtf/text/StringBuilderInternals.h:
(WTF::StringBuilder::allocateBuffer):
* Source/WTF/wtf/text/StringConcatenate.h:
* Source/WTF/wtf/text/StringConcatenateNumbers.h:
(WTF::StringTypeAdapter&lt;FormattedNumber&gt;::writeTo const):
(WTF::StringTypeAdapter&lt;FormattedCSSNumber&gt;::writeTo const):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::createInternal):
(WTF::StringImpl::foldCase):
(WTF::StringImpl::convertASCIICase):
(WTF::StringImpl::replace):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::copyCharacters):
(WTF::StringImpl::removeCharactersImpl):
(WTF::StringImpl::createByReplacingInCharacters):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::getCharacters8 const):
(WTF::StringView::getCharacters16 const):
(WTF::StringView::UpconvertedCharactersWithSize&lt;N&gt;::UpconvertedCharactersWithSize):
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::make8Bit):
(WTF::String::convertTo16Bit):
* Source/WTF/wtf/text/icu/UTextProviderLatin1.cpp:
(WTF::uTextLatin1Access):
(WTF::uTextLatin1Extract):
(WTF::textLatin1ContextAwareMoveInPrimaryContext):

Canonical link: <a href="https://commits.webkit.org/289466@main">https://commits.webkit.org/289466@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/14b34815d76ec4621d8573001e2083e2d5494053

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86970 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6479 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41322 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91828 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37707 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6756 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14547 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67225 "Failed to checkout and rebase branch from PR 39461") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24995 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89972 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5157 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78725 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47547 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4937 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33097 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36822 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/79754 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75432 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33976 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93714 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/85742 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14130 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/10352 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76030 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14334 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74653 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75227 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19558 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18060 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7009 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13558 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14149 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19440 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/108236 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13894 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26049 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17336 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15676 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -409,27 +409,30 @@ class StringImpl : private StringImplShape {
+    ALWAYS_INLINE static void copyCharacters(std::span<CharacterType> destination, std::span<const CharacterType> source)
     {
-        return copyElements(destination, source.data(), source.size());
+        RELEASE_ASSERT(destination.size() >= source.size());
```
Seems like this can go inside `copyElements` instead if we change it so it gets a destination size argument either separately or as part of moving it to use `span` arguments.

cdumez replied:
I agree, eventually. The thing is that this is a larger change (since there are more callers of `copyElements()`) and it is even more likely to negatively impact perf.

darinadler commented on Source/WTF/wtf/text/icu/UTextProviderLatin1.cpp:
```diff
@@ -144,7 +144,7 @@ static UBool uTextLatin1Access(UText* uText, int64_t index, UBool forward)
 
 WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
-    StringImpl::copyCharacters(const_cast<UChar*>(uText->chunkContents), unsafeMakeSpan(static_cast<const LChar*>(uText->context) + uText->chunkNativeStart, uText->chunkLength));
+    StringImpl::copyCharacters(unsafeMakeSpan(const_cast<UChar*>(uText->chunkContents), uText->chunkLength), unsafeMakeSpan(static_cast<const LChar*>(uText->context) + uText->chunkNativeStart, uText->chunkLength));
```
These kinds of getters that turn, say, `UText` data members into a span seem like they should be separate functions. They are the "boundary functions" that translate the unsafe API into our safer C++ idioms and I think they should contain as little else as possible.

darinadler commented on Source/WTF/wtf/text/icu/UTextProviderLatin1.cpp:
```diff
@@ -184,7 +184,7 @@ static int32_t uTextLatin1Extract(UText* uText, int64_t start, int64_t limit, UC
             trimmedLength = dest.size();
 WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
-        StringImpl::copyCharacters(dest.data(), unsafeMakeSpan(static_cast<const LChar*>(uText->context) + start, trimmedLength));
+        StringImpl::copyCharacters(dest, unsafeMakeSpan(static_cast<const LChar*>(uText->context) + start, trimmedLength));
```
If we had a function that takes a `UText` and returns a `span` then this would be using `subspan`.

cdumez replied:
I'm not familiar with ICU and UText is still a bit confusing to me. It seems UText has both `chunkContents` and `context`. Converting a UText to a span for `chunkContents` seems easy enough because I can rely on `chunkContents`  + `chunkLength` I think.

However, this line is about `context` which is a `void*` and I'm not sure what it corresponds to or its size.

cdumez replied:
Note that this code is still behind `WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN` and will have to be re-visited no matter what. I can use `std::span { }` instead of `unsafeMakeSpan()` for clarity.

darinadler replied:
Yes, `chunkContents` and `chunkLength` go hand in hand and point to the current chunk. So it’s great to have the `chunkSpan` function that takes a `UText`.

Since this is the implementation of a `UText` provider, `UTextProviderLatin1`. As the provider, we can store whatever pointer we want in `context`, and it’s effectively the pointer to “all the text” in a provider-specific sense. And so we should have a function specific to `UTextProviderLatin1` that makes a `span` that combines `context` with `a`. This also takes a `UText`, but needs to be a `UTextProviderLatin1`. Could assert `pFuncs == uTextLatin1Funcs`.

darinadler commented on Source/WTF/wtf/text/icu/UTextProviderLatin1.cpp:
```diff
@@ -183,9 +190,7 @@ static int32_t uTextLatin1Extract(UText* uText, int64_t start, int64_t limit, UC
-WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
-        StringImpl::copyCharacters(dest.data(), unsafeMakeSpan(static_cast<const LChar*>(uText->context) + start, trimmedLength));
-WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
+        StringImpl::copyCharacters(dest, latin1ContextSpan(uText).first(trimmedLength));
```
I think this should be:
```
StringImpl::copyCharacters(dest, latin1ContextSpan(uText).subspan(start, trimmedLength));
```

cdumez replied:
Oops. Indeed.

darinadler commented on Source/WTF/wtf/text/icu/UTextProviderLatin1.cpp:
```diff
@@ -301,7 +306,7 @@ static void textLatin1ContextAwareMoveInPrimaryContext(UText* text, int64_t nati
     text->chunkOffset = forward ? 0 : text->chunkLength;
 WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
-    StringImpl::copyCharacters(const_cast<UChar*>(text->chunkContents), unsafeMakeSpan(static_cast<const LChar*>(text->p) + (text->chunkNativeStart - text->b), text->chunkLength));
+    StringImpl::copyCharacters(chunkSpan(text), unsafeMakeSpan(static_cast<const LChar*>(text->p) + (text->chunkNativeStart - text->b), text->chunkLength));
```
This code can’t be right! I can’t find any code anywhere that is setting `text->p`. Clearly it’s not getting any worse with this pull request, but it must be dead code.

cdumez replied:
I left it as is for now as I'm not quite sure what to do about it. Via code auditing, it doesn't obviously look unused. there is code calling into it as far as I can see. However, when I run the layout tests, I don't see this code path getting exercised indeed..

cdumez committed:
Update StringImpl::copyCharacters() to take spans and do bounds validation
https://bugs.webkit.org/show_bug.cgi?id=286435

Reviewed by Darin Adler.

Also fix an issue in `JSRopeString::resolveToBuffer()` where we were artificially
reducing the size of the output buffer we were passing to `StringView::getCharacters()`
as it was tripping the subspan assertions. The code was safe since the buffer was large
enough, however, `StringView::getCharacters()` could not know that.

Note that the assertions are only debug assertions for now due to performance concerns
but we should try and make them release assertions wherever performance allows. The
assertions are currently in StringImpl::copyCharacters() but I intend to move them
to copyElements() in a follow-up as this requires even more refactoring.

* Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
(JSC::appendStringToData):
* Source/JavaScriptCore/runtime/StringConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::jsSpliceSubstrings):
* Source/WTF/wtf/HexNumber.h:
(WTF::StringTypeAdapter<HexNumberBuffer>::writeTo const):
* Source/WTF/wtf/URLHelpers.cpp:
(WTF::URLHelpers::escapeUnsafeCharacters):
* Source/WTF/wtf/text/StringBuilder.cpp:
(WTF::StringBuilder::append):
* Source/WTF/wtf/text/StringBuilderInternals.h:
(WTF::StringBuilder::allocateBuffer):
* Source/WTF/wtf/text/StringConcatenate.h:
* Source/WTF/wtf/text/StringConcatenateNumbers.h:
(WTF::StringTypeAdapter<FormattedNumber>::writeTo const):
(WTF::StringTypeAdapter<FormattedCSSNumber>::writeTo const):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::createInternal):
(WTF::StringImpl::foldCase):
(WTF::StringImpl::convertASCIICase):
(WTF::StringImpl::replace):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::copyCharacters):
(WTF::StringImpl::removeCharactersImpl):
(WTF::StringImpl::createByReplacingInCharacters):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::getCharacters8 const):
(WTF::StringView::getCharacters16 const):
(WTF::StringView::UpconvertedCharactersWithSize<N>::UpconvertedCharactersWithSize):
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::make8Bit):
(WTF::String::convertTo16Bit):
* Source/WTF/wtf/text/icu/UTextProviderLatin1.cpp:
(WTF::uTextLatin1Access):
(WTF::uTextLatin1Extract):
(WTF::textLatin1ContextAwareMoveInPrimaryContext):

Canonical link: https://commits.webkit.org/289466@main
